lib/spack/spack/solver/asp.py

```diff
@@ -1439,7 +1475,11 @@ def condition(
 
             condition_id = next(self._id_counter)
             trigger_id = self._get_condition_id(
-                required_spec, cache=self._trigger_cache, body=True, transform=transform_required
+                required_spec,
+                cache=self._trigger_cache,
+                body=True,
+                transform=transform_required,
+                source=source,
```

Comment:
The `transform_*` functions add/remove clauses vs. modifying them (`_spec_clauses` needs to attach the `source` to generated `node_flag` attributes); overall, I would consider it confusing to bundle any of the currently generated clauses related to node flags into a callback.

Reply:
> The transform_* functions add/remove clauses vs. modifying them

Who says that? Also, what's the difference? :slightly_smiling_face:  The `transform_` function takes a list of facts, does what it needs to do, and returns the final list. I added that callback to avoid having a long list of arguments, each one used in a very specific place.